File: benchmarks/fixtures/webapp_csharp/Services/AuthenticationService.cs
namespace Webapp.Services;

using System.Collections.Generic;
using Webapp.Auth;
using Webapp.Database;
using Webapp.Models;
using Webapp.Util;

/// <summary>
/// Orchestrates the full authentication workflow.
///
/// Deep call chain entry point:
/// Authenticate() -> Login() -> GenerateToken() -> ExecuteQuery() -> GetConnection()
/// </summary>
public class AuthenticationService : BaseService
{
    private static readonly Logger Log = Logger.GetLogger("services.authentication");

    private readonly AuthService _authService;
    private readonly DatabaseConnection _db;

    public AuthenticationService(DatabaseConnection db) : base("authentication")
    {
        Log.Info("Creating AuthenticationService");
        _authService = new AuthService(new TokenService());
        _db = db;
    }

    /// <summary>
    /// Perform the full authentication flow.
    /// </summary>
    public string Authenticate(string email, string password)
    {
        RequireInitialized();
        Log.Info($"Authenticating user: {email}");

        string token = _authService.Login(email, password);

        var session = new Dictionary<string, object>
        {
            ["token"] = token,
            ["email"] = email
        };
        _db.Insert("sessions", session);

        Log.Info($"Authentication successful for: {email}");
        return token;
    }

    public void Logout(string token)
    {
        Log.Info("Logging out");
        _authService.Logout(token);
    }

    public User GetCurrentUser(string token)
    {
        Log.Info("Getting current user");
        return _authService.GetCurrentUser(token);
    }
}

File: benchmarks/fixtures/webapp_cpp/services/authentication_service.cpp
#include "services/authentication_service.h"

#include <string>

#include "database/database_connection.h"
#include "models/user.h"
#include "util/logger.h"

namespace webapp {
namespace services {

AuthenticationService::AuthenticationService(database::DatabaseConnection& db)
    : BaseService("authentication"),
      log_(util::Logger::get_logger("services.authentication")),
      db_(db),
      token_service_(db),
      auth_service_(token_service_) {
    log_.info("Creating AuthenticationService");
}

/// Perform the full authentication flow. Hop 2 of the deep call chain.
std::string AuthenticationService::authenticate(const std::string& email,
                                               const std::string& password) {
    require_initialized();
    log_.info("Authenticating user: " + email);

    std::string token = auth_service_.login(email, password);

    database::Row session;
    session["token"] = token;
    session["email"] = email;
    db_.insert("sessions", session);

    log_.info("Authentication successful for: " + email);
    return token;
}

/// Tear down the caller's session.
void AuthenticationService::logout(const std::string& token) {
    log_.info("Logging out");
    auth_service_.logout(token);
}

/// Resolve the user behind a token.
models::User AuthenticationService::get_current_user(const std::string& token) {
    log_.info("Getting current user");
    return auth_service_.get_current_user(token);
}

}  // namespace services
}  // namespace webapp
